### PR TITLE
docker_container: simplify minimal required version per option handling

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
docker_container supports some options which require a newer docker/docker-py version and/or a newer docker API version. Currently, checking whether this is done correctly is done manually with some feature flags and comparisons. This PR attempts to improve on this situation by allowing to specify minimal versions with a minimal amount of effort (that's implemented in the first commit), and by automating the whole "can this option be passed here or not" game (not yet done).

@akshay196 this is what I meant in my comment in #36831.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
docker_container

##### ANSIBLE VERSION
```
2.8.0
```

### CODE REVIEW

**Comment by felixfontein** [comment]
Tests should stop failing when #47712 is merged.

File: lib/ansible/modules/cloud/docker/docker_container.py [hunk 1]
```diff
@@ -2527,34 +2528,77 @@ def _parse_comparisons(self):
...
-            self.fail("docker or docker-py version is %s. Minimum version required is 1.9 to set blkio_weight option.")
+        self.option_minimal_versions = dict(
+            init=dict(docker_py_version='2.2.0', docker_api_version='1.25'),
+            uts=dict(docker_py_version='3.5.0'),
+            blkio_weight=dict(docker_py_version='1.9.0'),
+            cpuset_mems=dict(docker_py_version='2.3.0'),
+            ipvX_address_supported=dict(docker_py_version='1.9.0', detect_usage=detect_ipvX_address_usage,
+                                        usage_msg='ipv4_address or ipv6_address in networks'),
+            runtime=dict(docker_py_version='2.4.0', docker_api_version='1.12'),
+            healthcheck=dict(docker_py_version='2.0.0'),
```
Comment (akshay196) [inline lib/ansible/modules/cloud/docker/docker_container.py]: @felixfontein I just check [docker api changes](https://docs.docker.com/engine/api/version-history/#v125-api-changes), **healthcheck** requires minimum docker api version is `1.25`. I missed condition for minimum docker api version. Could you please add it now?

Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_container.py]: Done! I think we need to go through all existing options anyway to make sure we got the correct minimal API version and minimal docker-py version for all options. I wouldn't be surprised if some of the current checks are not completely correct, or haven't been there in the first place.

**Comment by felixfontein** [comment]
Rebased to account for #47712 being merged.

**Comment by felixfontein** [comment]
There was a strange spurious test failure; [link to test](https://app.shippable.com/github/ansible/ansible/runs/90797/49/tests). Again, somewhen later: [link to test](https://app.shippable.com/github/ansible/ansible/runs/91137/52/tests).

**Comment by felixfontein** [comment]
I think this is done now. I hope I didn't introduce any bugs, though :)

**Review by kassiansun — APPROVED** [review]
shipit

**Review by akshay196 — COMMENTED** [review]
Just minor corrections.

File: lib/ansible/modules/cloud/docker/docker_container.py [hunk 2]
```diff
@@ -2524,37 +2519,98 @@ def _parse_comparisons(self):
...
+            if 'docker_api_version' in data:
+                support_docker_api = self.docker_api_version >= LooseVersion(data['docker_api_version'])
+            data['supported'] = support_docker_py and support_docker_api
+            # Fail if option is not supported but used
+            if not data['supported']:
+                # Test whether option is specified
+                if 'detect_usage' in data:
+                    used = data['detect_usage']()
+                else:
+                    used = self.module.params.get('option') is not None
```
Comment (akshay196) [inline lib/ansible/modules/cloud/docker/docker_container.py]:
```suggestion
                    used = self.module.params.get(option) is not None
```

File: lib/ansible/modules/cloud/docker/docker_container.py [hunk 3]
```diff
@@ -2524,37 +2519,98 @@ def _parse_comparisons(self):
...
+                    if used and 'default' in self.module.argument_spec[option]:
+                        used = self.module.params['option'] != self.module.argument_spec[option]['default']
```
Comment (akshay196) [inline lib/ansible/modules/cloud/docker/docker_container.py]:
```suggestion
                        used = self.module.params[option] != self.module.argument_spec[option]['default']
```

File: lib/ansible/modules/cloud/docker/docker_container.py (hunk 2 above)
Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_container.py]: Haha, that's a stupid one :D Thanks for noticing!

**Comment by felixfontein** [comment]
I'll do some more testing (and I'm adjusting the tests to also work with older docker-py versions which don't support everything yet).

**Comment by felixfontein** [comment]
Ok, the tests now work with any (or at least a lot :D ) docker-py versions. I've ran the tests with docker-py 1.8.0, 1.9.0, 1.10.0, 1.10.6, 2.0.0, 2.1.0, 2.2.0, 2.3.0, 2.4.0, 2.5.0, 2.6.0, 2.7.0, 3.0.0, 3.1.0, 3.2.0, 3.3.0, 3.4.0, 3.5.0, and 3.5.1 (latest). I guess it should be really OK now :-)

I'll move the docker_py_version / docker_api_version part of testing to setup_docker, that will be useful also in other docker_* tests. After that it should be ready to go.

**Comment by felixfontein** [comment]
Should be done now!

**Comment by akshaygin** [comment]
LGTM

**Comment by felixfontein** [comment]
Let's get #47814 merged first (and then rebase this).

**Comment by felixfontein** [comment]
Squashing to make rebase easier.

**Comment by felixfontein** [comment]
Tested with docker-py 1.8.0, 1.9.0, 1.10.0, 2.0.0 and 3.5.1 (latest). Still works :) Now it's *really* ready to go ;)

**Comment by kassiansun** [comment]
shipit

**Comment by willthames** [comment]
Merged, thanks @felixfontein for the fix and @kassiansun for the review

**Comment by felixfontein** [comment]
I'll create a stable-2.6 backport once some of the existing ones are merged. Otherwise it is a PITA and will generate a *lot* of merge conflicts...

**Comment by felixfontein** [comment]
Backport included in #47288.

**Pull request merged**